### PR TITLE
Add Documentation to client/domain folder

### PR DESCRIPTION
**Pull request by dchen97**
* Added comments according to Google's JavaScript Style Guide to describe classes, variables, and functions for all classes in the `client/domain` folder.
* Added short development guide
* Referenced development guide in CONTRIBUTING.md

### CODE REVIEW

File: client/domain/CodeSubmissionObjectFactory.js [hunk 1]
```diff
@@ -60,6 +126,13 @@ tie.factory('CodeSubmissionObjectFactory', [
       this._preprocessedCodeLines = replacementCodeLines;
     };
 
+    /**
+     * Splits and appends the string in the codeToAppend param to the Array in
+     * _preprocessedCodeLines property.
+     *
+     * @param codeToAppend    should be a string with the code to
```
Comment (theresa-lee) [inline client/domain/CodeSubmissionObjectFactory.js]: extra spaces

File: client/domain/CodeSubmissionObjectFactory.js [hunk 2]
```diff
@@ -69,6 +142,13 @@ tie.factory('CodeSubmissionObjectFactory', [
       }
     };
 
+    /**
+     * Splits and prepends the string in the codeToPrepend param to the Array in
+     * _preprocessedCodeLines property.
+     *
+     * @param codeToPrepend   should be a string with the code to
```
Comment (theresa-lee) [inline client/domain/CodeSubmissionObjectFactory.js]: extra spaces

File: client/domain/CodeSubmissionObjectFactory.js [hunk 3]
```diff
@@ -96,6 +178,12 @@ tie.factory('CodeSubmissionObjectFactory', [
     };
 
     // Static class methods.
+    /**
+     * Returns a CodeSubmission object built from the rawCode param
+     *
+     * @param rawCode     should be a string with the unprocessed student code
```
Comment (theresa-lee) [inline client/domain/CodeSubmissionObjectFactory.js]: extra spaces

File: client/domain/CodeEvalResultObjectFactory.js [hunk 4]
```diff
@@ -19,59 +19,179 @@
...
   function() {
+    /**
+     * CodeEvalResult stores all of the results for each test and any related
+     * errors associated with a given code submission.
+     */
+
+    /**
+     * Constructor for CodeEvalResult
+     *
+     * @param {string} preprocessedCode    Unprocessed student code
```
Comment (dchen97) [inline client/domain/CodeEvalResultObjectFactory.js]: Done!

File: client/domain/CodeEvalResultObjectFactory.js [hunk 5]
```diff
@@ -19,59 +19,179 @@
...
+     * @returns {Object}
+     */
     CodeEvalResult.prototype.getLastTaskResults = function() {
       return this._correctnessTestResults[
         this._correctnessTestResults.length - 1];
     };
 
+    /**
+     * A getter for the _buggyOutputTestResults
+     * property. The function should return a list of Objects where results
```
Comment (dchen97) [inline client/domain/CodeEvalResultObjectFactory.js]: Done!

File: client/domain/CodeEvalResultObjectFactory.js [hunk 6]
```diff
@@ -19,59 +19,179 @@
...
     };
 
     // Static class methods.
+    /**
+     * This method creates and returns a CodeEvalResult object from the params
+     * specified.
+     *
+     * @param {string} preprocessedCode preprocessed submitted code
+     * @param {string} output output from the code
+     * @param {Array} correctnessTestResults  correctness test results
```
Comment (dchen97) [inline client/domain/CodeEvalResultObjectFactory.js]: Done!

File: DevDoc.md [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+
+* `scripts`: This folder has all of our scripts - mainly used for testing purposes.
+
+* `third_party`: This folder houses all of our approved third party code.
+
+## Testing
+
+All of our unit tests are created in our `*Spec.js`-named files. The `*` represents the file name for the component(s) that the sister `Spec` file is testing.
+
+It is highly encouraged that you write tests for as much code coverage as possible. In order to test for coverage, you can run `bash scripts/run_karma_tests.sh --enable-coverage` to run Karma tests in addition to generating coverage reports
```
Comment (dchen97) [inline DevDoc.md]: Done!

File: client/domain/BuggyOutputTestObjectFactory.js [hunk 8]
```diff
@@ -19,21 +19,62 @@
...
+     */
     BuggyOutputTest.prototype.getBuggyFunctionName = function() {
       return this._buggyFunctionName;
     };
 
+    /**
+     * Returns the list of messages that a buggy function has
+     *    attached to it
+     *
+     * @returns {Array}   should be array of strings
```
Comment (dchen97) [inline client/domain/BuggyOutputTestObjectFactory.js]: Done!

File: client/domain/CodeSubmissionObjectFactory.js [hunk 9]
```diff
@@ -19,39 +19,105 @@
...
+     * the user submitted.
+     *
+     * @returns {string} rawCode
+     */
     CodeSubmission.prototype.getRawCode = function() {
       return this._rawCode;
     };
 
+    /**
+     * Returns the list of preprocessedCodeLines joined together to be one
```
Comment (dchen97) [inline client/domain/CodeSubmissionObjectFactory.js]: Done!

File: client/domain/CodeSubmissionObjectFactory.js [hunk 10]
```diff
@@ -19,39 +19,105 @@
...
     CodeSubmission.prototype.getRawCodeLineIndexes = function() {
       return this._rawCodeLineIndexes;
     };
 
-    // Replaces the entire code. The line numbers should not change.
+    /**
+     * Replaces the entirety of the code in the _preprocessedCodeLines property
+     * with those of the replacementCode param.
+     *
+     * @param replacementCode   should be an Array of strings that is
```
Comment (dchen97) [inline client/domain/CodeSubmissionObjectFactory.js]: Done

File: client/domain/CodeSubmissionObjectFactory.js (hunk 1 above)
Comment (dchen97) [inline client/domain/CodeSubmissionObjectFactory.js]: Done!

File: client/domain/CodeSubmissionObjectFactory.js (hunk 2 above)
Comment (dchen97) [inline client/domain/CodeSubmissionObjectFactory.js]: Done!

File: client/domain/CodeSubmissionObjectFactory.js (hunk 3 above)
Comment (dchen97) [inline client/domain/CodeSubmissionObjectFactory.js]: Done!

File: client/domain/ErrorTracebackObjectFactory.js [hunk 11]
```diff
@@ -20,21 +20,59 @@
...
+     * coordinates for a given Error in a student's submission.
+     */
+
+    /**
+     * Constructor for ErrorTraceback
+     *
+     * @param {string} errorMessage string describing the error
+     * @param {Array} tracebackCoordinates Array of ErrorTracebackCoordinates
+     *    objects describing the sources of
+     *    each error
```
Comment (dchen97) [inline client/domain/ErrorTracebackObjectFactory.js]: Done!

File: client/domain/ErrorTracebackObjectFactory.js [hunk 12]
```diff
@@ -20,21 +20,59 @@
...
+       * @type {string}
+       * @private
+       * */
       this._errorMessage = errorMessage;
-      // A list of ErrorTracebackCoordinates objects. Each object contains a
-      // lineNumber and columnNumber (both 1-indexed), representing the
-      // locations described in the error trace.
+
+      /**
+       * // A list of ErrorTracebackCoordinates objects. Each object contains a
```
Comment (dchen97) [inline client/domain/ErrorTracebackObjectFactory.js]: Done!

File: client/domain/FeedbackObjectFactory.js [hunk 13]
```diff
@@ -20,29 +20,87 @@
...
+     * or not.
+     *
+     * @returns {boolean}
+     */
     Feedback.prototype.isAnswerCorrect = function() {
       return this._answerIsCorrect;
     };
 
+    /**
+     * Appends a FeedbackParagraph is of type text to the _paragraphs Array.
```
Comment (dchen97) [inline client/domain/FeedbackObjectFactory.js]: Done!

**Pull request merged**